### PR TITLE
feat(weapons): implement autonomous ZMQ inference worker

### LINKED ISSUE #35: feat(weapons): implement autonomous ZMQ inference worker for E2E telemetry
# feat(weapons): implement autonomous ZMQ inference worker for E2E telemetry

## Executive Summary & Context
To achieve End-to-End (E2E) system validation, we must deploy the Weapon Detection microservice (`src/modules/weapons/inference.py`). We are adopting a "Tracer Bullet" development methodology: our immediate priority is establishing a robust, fault-tolerant Inter-Process Communication (IPC) pipeline from the edge-node camera to the Orchestrator sink. 

Furthermore, to support non-destructive, vector-based rendering on the future React Frontend, this worker must extract and transmit absolute Bounding Box coordinates. This architectural decoupling ensures the UI can draw dynamic DOM overlays without permanently mutating or burning pixels into the raw evidence frames.

## Architectural Directives & Network Contracts

The worker must act as an autonomous node in the Fan-Out topology and adhere strictly to the following constraints:

**1. Memory Safety & Throttling (Chesterton's Fence):**
The worker must subscribe to the video stream via `zmq.SUB`. Crucially, the socket MUST be configured with `socket.setsockopt(zmq.CONFLATE, 1)`. If inference time exceeds the 10 FPS ingestion rate, the worker must explicitly drop stale frames rather than queuing them. Failure to implement this will result in fatal out-of-memory (OOM) operating system kills.

**2. Cold-Start Mitigation:**
PyTorch/CUDA environments suffer from high latency during the initial tensor allocation. The worker must execute a "warmup" pass using a dummy zero-tensor during bootstrapping to prevent processing delays when the first real threat is captured.

**3. IPC Payload Contract (Updated for Vector UI):**
The worker must exclusively push (`zmq.PUSH`) data to the Orchestrator (`tcp://127.0.0.1:5556`) when a threat exceeds the confidence threshold (>= 0.50). Silent frames should generate zero network traffic. The payload must include the `bbox` array `[x_min, y_min, x_max, y_max]` cast as integers.

```json
{
  "module": "weapons",
  "camera_id": "main_camera",
  "timestamp": "2026-04-12T15:00:00.123456",
  "frame_data": "<base64_encoded_jpeg>", 
  "detections": [
    {
      "class": "pistol", 
      "confidence": 0.88,
      "bbox": [150, 200, 350, 400]
    }
  ]
}
```

## Implementation Steps
1. Populate `src/modules/weapons/inference.py` with the approved ZMQ worker boilerplate.
2. Extract the spatial coordinates from the YOLO tensor (`box.xyxy[0].cpu().numpy().astype(int).tolist()`) and append them to the detection dictionary to satisfy the UI contract.
3. Link the `MODEL_WEIGHTS` constant to the absolute path of the currently available `best.pt` file generated from the initial Roboflow experiments.
4. Integrate the worker's execution entry point into the `processes` array within `main.py`.

## Acknowledged Technical Debt
The current model weights integrated via this issue are baseline iterations and possess known performance degradation regarding small/distant objects. This is documented and accepted. A subsequent Epic will address SOTA Fine-Tuning using CCTV-specific datasets. Once that training is complete, the new weights will be hot-swapped into this pipeline without requiring code modifications.

## Architectural Decision: YAGNI & Stateless Domain
Reviewers note: Unlike the `face` module (which is stateful and requires database lookups/embeddings isolated in `services/face_processor.py`), weapon detection is a pure, stateless I/O function. Applying YAGNI (You Aren't Gonna Need It), we intentionally bypassed creating a redundant `services/weapon_processor.py`. All parsing logic resides within the edge worker to prevent over-engineering and eliminate unnecessary call-stack latency.

## Definition of Done (DoD)
- [ ] The weapon worker process boots successfully and reserves its VRAM allocation without crashing.
- [ ] The worker accurately drops frames (via `zmq.CONFLATE`) when intentionally throttled, demonstrating immunity to buffer bloat.
- [ ] Spatial coordinates (`bbox`) are successfully extracted from the PyTorch tensor and formatted as an integer array.
- [ ] Presenting a weapon to the camera successfully generates the strict JSON payload (including `bbox`) and transmits it to the Orchestrator.

### PR DESCRIPTION
**Pull request by Armando2030**
### Summary
Implements the weapon detection inference worker as an autonomous ZeroMQ subscriber in the Fan-Out topology. The worker subscribes to the video stream, runs YOLOv11 inference on each frame, and pushes threat detections (knife/pistol) to the Orchestrator for rule evaluation.

### Changes

**New files:**
- `src/modules/weapons/inference.py` — Autonomous ZMQ worker for weapon detection using YOLO

**Modified files:**
- `main.py` — Added Weapon_Process to the multiprocessing array (now 4 processes)

### Architecture

```
                        [Camera]
                           │
                     [stream.py]
                      PUB :5555
                           │
            ┌──────────────┼──────────────┐
            │              │              │
         [face]       [weapons]       [pose]
        SUB:5555      SUB:5555       SUB:5555
            │              │              │
            └──────────────┼──────────────┘
                           │
                      PUSH → PULL
                           │
                      [rules.py]
                     Orchestrator
                       :5556
```

### Implementation details
- **Model**: YOLOv10 with custom weights (`models_testing/Object Detection/best.pt`)
- **Classes detected**: knife, pistol (skips generic "object" class)
- **Confidence threshold**: >= 0.50
- **Frame dropping**: zmq.CONFLATE=1 ensures only the newest frame is processed
- **Warmup pass**: Dummy tensor inference on startup to avoid cold-start latency
- **Evidence capture**: Detected frames are JPEG-encoded (quality 75) and base64-encoded for MinIO storage
- **Payload format**: Includes class, confidence, bbox [x_min, y_min, x_max, y_max], and frame_data

### How to test
1. `docker compose down -v && docker compose up`
2. Enroll a face via the API (POST /api/persons/ + curl enroll)
3. `cd backend && source venv/bin/activate && python main.py`
4. System boots 4 processes: Orchestrator, Face, Weapons, Ingestion
5. Stand in front of the webcam — should see [ACCESS GRANTED]
6. Hold an image of a weapon in front of the webcam — should see [WEAPON DETECTED] and the Orchestrator creates an incident + alert in PostgreSQL

### Dependencies
- `ultralytics` (already in requirements.txt)
- `best.pt` model weights (already in models_testing/Object Detection/)

### CODE REVIEW

**Review by mauriziokeningen — APPROVED** [review]
### PR Audit & E2E Integration Report

The branch has been thoroughly audited and tested on local hardware (RTX 4070 Ti Super). The following enhancements and fixes were implemented to ensure production readiness:

#### 1. Dependency & Environment Stability
* **Core Fix:** Upgraded `ultralytics` to `8.4.37`. This resolves the security restriction in PyTorch 2.6 that was preventing model initialization.
* **Hardware Agnostic:** Verified the "opt-in" GPU workflow. The codebase remains compatible with CPU-only environments for the rest of the team while supporting CUDA acceleration where available.

#### 2. Storage Pipeline & Data Integrity
* **Data Contract:** Patched `rules.py` to handle Base64 encoded frames from the AI workers. Added explicit `base64.b64decode` logic to ensure raw bytes reach the S3 client.
* **Infrastructure Alignment:** Synchronized bucket naming conventions. The system now correctly targets the `incident-evidence` bucket, resolving previous `NoSuchBucket` errors.

#### 3. Documentation & Observability
* **Tech Debt Preservation:** Restored critical architectural documentation regarding synchronous network I/O in the Orchestrator. This is vital for the V2 roadmap (Celery/Redis integration).
* **Diagnostic Tooling:** Merged `check_system.py`. This E2E script provides a unified health check for PostgreSQL (pgvector) and MinIO, handling recursive object listing and metadata validation.

#### 4. Final Validation Result
* **Face Detection:** Operational.
* **Weapon Detection:** Operational (Pistol/Knife classes verified).
* **Persistence:** Successfully verified 6+ incidents with physical `.jpg` evidence stored in MinIO.

**Verdict:** System is stable, documented, and auditable.
